Support Transfer coin and pays with same coin

lxfind opened the pull request:
This PR adds the ability to transfer a coin object and also pays gas with the same object.
Specifically it does the following:
1. In wallet, when gas is not specified, it will automatically pay with the transfer object, instead of trying to pick a gas object. We will need to update docs to reflect this.
2. The object_ref field in TransferCoin transaction kind is now option, and when it's none, we will be transferring the gas object.
3. Make sure that this type of transfer can not happen in a batch transaction
4. Added the ability to deduct and change owner of the gas object at the same time
5. Added a test.
6. Updated docs.

sblackshear commented:
Nice!!

One clarifying question: I think your previous proposal also included an optional `--amount <int_amount>` argument that will chip a piece off the coin in question rather than sending the whole thing.  Is that still planned for the future?

lxfind commented:
Yes, plan to add amount separately

kchalkias commented on sui_core/src/execution_engine.rs:
```diff
@@ -90,13 +91,18 @@ fn execute_transaction<S: BackingPackageStore>(
-                transfer(temporary_store, object, recipient)
+                match &object_ref {
+                    Some((object_id, _, _)) => {
+                        // unwrap is is safe because we built the object map from the transactions
```
double "is"

kchalkias commented on sui_core/src/transaction_input_checker.rs:
```diff
@@ -131,17 +131,25 @@ where
+                }
+                None => {
+                    fp_ensure!(
+                        !transaction.kind.is_batch_transaction(),
```
+1

666lcz commented on sui_types/src/messages.rs:
```diff
@@ -293,11 +253,25 @@ where
     }
 
+    /// Create a native coin transfer transaction that pays gas using the transfer coin.
+    pub fn new_transfer_gas(
```
I am now a bit confused about the `TransferCoin` method now, does it only allow transferring `Coin<SUI>` or any `Coin<T>`? It looks like it's the former based on this comment(because you cannot pay gas with arbitrary Coin<T>) and if so, should we rename `TransferCoin` to `TransferSuiCoin` or extend the capability to support `Coin<T>`?

lxfind replied:
Oops you are right! Let me rethink

sblackshear replied:
Good point!

Another semi-related usability issue perhaps worth considering at the same time. This API lets you do "send a SUI coin worth `N - gas_price` to `recipient`". It's somewhat hard to imagine a scenario where you are indifferent about the amount that `recipient` gets, or have perfectly created a coin whose value is `amount + gas_price`. That makes me think we maybe only want this API when `--amount` is passed...

kchalkias replied:
Another qq: would that work with other contracts that include gas transfer inside their logic? i.e., https://github.com/MystenLabs/sui/blob/main/sui_programmability/examples/nfts/sources/DiscountCoupon.move#L49

lxfind replied:
I actually think there are legit cases where you are indifferent about the amount that recipient gets, especially if we assume our gas will be so cheap that it shouldn't be noticed. For instance, I often want to consolidate my accounts from different wallet/exchange, but today in Eth world you always leave dust behind which is very very annoying.

sblackshear replied:
Fair point re: the case when you at a different address is the recipient.

666lcz commented:
Follow up on @sblackshear 's question on `--amount <int_amount>`, will there be any difference in terms of gas consumption between using the magic transfer method(using the same coin) and transferring the whole object and paying with a separate gas object? This will have a big implication on wallet implementation, which will try to save gas for users.

Clay-Mysten reviewed (approved):
Thanks, Chris. wallet.md changes LGTM.

lxfind commented:
> How can we generalize this logic more cleanly to allow for other features which also seek to circumvent this rule?

To get magic coin merge and split work, I think there is some more work to be done first:
Currently coin merge and split is done through Move call. But to be able to share gas object, we may need to first convert them to not use Move call to achieve coin merge and split? Or otherwise we need to introduce a mechanism that could deduct gas using one of the objects in the Move call input. 
Then there is also a similar issue where we will need a general coin merge/split, and also a SUI-specific coin merge/split that supports magic gas.
So I am not quite sure at this point how far I can go in terms of making this mechanism fully generic. Need to think about it.

oxade commented:
@lxfind the current basic gas selection logic forbids the gas object used for payment from being in the transaction itself
See here https://github.com/MystenLabs/sui/blob/3746dbbe5ce1cb698e0118454431cbd8266afa85/sui_core/src/gateway_state.rs#L610
Since this is now different, you might want to make sure these logics don't collide. Also might wanna change that comment.

lxfind commented:
> @lxfind the current basic gas selection logic forbids the gas object used for payment from being in the transaction itself See here
> 
> https://github.com/MystenLabs/sui/blob/3746dbbe5ce1cb698e0118454431cbd8266afa85/sui_core/src/gateway_state.rs#L610
> 
> 
> Since this is now different, you might want to make sure these logics don't collide. Also might wanna change that comment.

Does this matter though? When we use magic transfer, the gas selection logic won't come into play (because we don't need to select a gas anymore).
